src/projections.jl

```diff
@@ -0,0 +1,248 @@
+
+projection of vector `z` on zero cone i.e. K = {0} or its dual
+"""
+function projection_on_set(::DefaultDistance, ::MOI.Zeros, z::Array{Float64}, dual=false)
+    return dual ? z : zeros(Float64, size(z))
+end
+
+function projection_on_set(::DefaultDistance, set::MOI.EqualTo, z::Float64, dual=false)
+    return dual ? z : set.value  # i doubt this value
+end
+
+"""
+    projection_on_set(::DefaultDistance, ::MOI.Nonnegatives, z::Array{Float64})
+
+projection of vector `z` on Nonnegative cone i.e. K = R+
+"""
+function projection_on_set(::DefaultDistance, ::MOI.Nonnegatives, z::Array{Float64})
+    return max.(z, 0.0)
+end
+
+"""
+    projection_on_set(::NormedEpigraphDistance{p}, ::MOI.SecondOrderCone, z::Array{Float64})
+
+projection of vector `z` on second order cone i.e. K = {(t, x) ∈ R+ × Rn |  ||x|| ≤ t }
+"""
+function projection_on_set(::NormedEpigraphDistance{p}, ::MOI.SecondOrderCone, z::Array{Float64}) where {p}
+    t = z[1]
+    x = z[2:length(z)]
+    norm_x = LinearAlgebra.norm(x, p)
+    if norm_x <= t
+        return copy(z)
+    elseif norm_x <= -t
+        return zeros(Float64, size(z))
+    else
+        result = zeros(Float64, size(z))
+        result[1] = 1.0
+        result[2:length(z)] = x / norm_x
+        result *= (norm_x + t) / 2.0
+        return result
+    end
+end
+
+function projection_on_set(::DefaultDistance, cone::MOI.SecondOrderCone, z::Array{Float64})
+    return projection_on_set(NormedEpigraphDistance{2}(), cone, z)
+end
+
+"""
+    projection_on_set(::DefaultDistance, ::MOI.PositiveSemidefiniteConeTriangle, z::Array{Float64})
+    
+projection of vector `z` on positive semidefinite cone i.e. K = S^n⨥
+"""
+function projection_on_set(::DefaultDistance, ::MOI.PositiveSemidefiniteConeTriangle, z::Array{Float64})
+    dim = Int64(floor(√(2*length(z))))
+    X = unvec_symm(z, dim)
+    λ, U = LinearAlgebra.eigen(X)
+    D = LinearAlgebra.Diagonal(max.(λ, 0))
+    return vec_symm(U * D * U')
+end
+
+"""
+    unvec_symm(x, dim)
+
+Returns a dim-by-dim symmetric matrix corresponding to `x`.
+
+`x` is a vector of length dim*(dim + 1)/2, corresponding to a symmetric
+matrix; the correspondence is as in SCS.
+X = [ X11 X12 ... X1k
+        X21 X22 ... X2k
+        ...
+        Xk1 Xk2 ... Xkk ],
+where
+vec(X) = (X11, sqrt(2)*X21, ..., sqrt(2)*Xk1, X22, sqrt(2)*X32, ..., Xkk)
+"""
+function unvec_symm(x, dim)
+    X = zeros(dim, dim)
+    for i in 1:dim
+        for j in i:dim
+            @inbounds X[j,i] = X[i,j] = x[(i-1)*dim-Int(((i-1)*i)/2)+j]
+        end
+    end
+    # X = X + X'
+    X /= √2
+    for i in 1:dim
+        X[i,i] *= √2
+    end
+    return X
+end
+
+"""
+    vec_symm(X)
+
+Returns a vectorized representation of a symmetric matrix `X`.
+Vectorization (including scaling) as per SCS.
+`vec(X) = (X11, sqrt(2)*X21, ..., sqrt(2)*Xk1, X22, sqrt(2)*X32, ..., Xkk)`
+"""
+function vec_symm(X)
+    X = copy(X)
+    X *= √2
```

Comment:
If you do this, there is no need to `copy` `X` as it gets rewritten as `X = X * sqrt(2)`

Reply:
removed